### ISSUE TITLE
`useSpeechRecognition` functions `listen` and `stop` are regenerated every time `listening` or `supported` changes

### ISSUE THREAD

**Issue by maxwellhaydn** [MikeyParton/react-speech-kit#31]
With `useSpeechRecognition`, every time one of the state values `listening` or `supported` changes, it triggers a re-render. With each render, a new version of the `listen` and `stop` functions is generated. This is inefficient and can lead to surprising behavior. For example:

```
import React from 'react';
import ReactDOM from 'react-dom';
import { useSpeechRecognition } from 'react-speech-kit';

const App = (props) => {
    const { listen, listening, stop } = useSpeechRecognition({
        onResult: (result) => console.log(result)
    });

    const handleClick = () => {
        setTimeout(() => {
            console.log('stop listening');
            stop();
        }, 1000);

        listen();
    };

    return (
        <div className="app">
            <button onClick={handleClick}>{listening ? 'Listening' : 'Speak'}</button>
        </div>
    );
};

ReactDOM.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
  document.getElementById('root')
);
```

I would expect this to stop listening one second after clicking the button, but it continues listening indefinitely. When I run this through the Chrome debugger, I can see that `stop` is called, but it simply returns without doing anything because `listening` is false:

```
  const stop = () => {
    if (!listening || !supported) return;
    ...
```

This happens because of a combination of two closures:

1. `stop`: Inside `stop`, `listening` always has the value it had when `stop` was first defined
2. The`setTimeout` callback: Inside the callback, `stop` always has the value it had when the `setTimeout` callback was first defined

When you click the button, the `listen` function calls `setListening(true)`, triggering a re-render. This generates a new version of the `stop` callback where `listening` is true. But the `setTimeout` callback still holds a reference to the original version of `stop` where `listening` was false, and that is the version that gets called after the timeout elapses. See [Each Render Has Its Own Props and State](https://overreacted.io/a-complete-guide-to-useeffect/#each-render-has-its-own-props-and-state) and [Each Render Has Its Own Event Handlers](https://overreacted.io/a-complete-guide-to-useeffect/#each-render-has-its-own-event-handlers) in Dan Abramov's "A Complete Guide to useEffect" for a good explanation of this.

The issue doesn't happen with the example in the README because there's no second closure there. When you mouse down on the button, the `listen` function calls `setListening(true)`, triggering a re-render and generating a new version of the `stop` callback where `listening` is true. This new callback is then attached to the button, so when you mouse up, you call the new callback and everything works.

This issue is a permutation of [Why am I seeing stale props or state inside my function?](https://reactjs.org/docs/hooks-faq.html#why-am-i-seeing-stale-props-or-state-inside-my-function) in the Hooks API FAQ. The behavior in the FAQ makes sense to me, but I wouldn't expect the same issue with `listen` and `stop` from `useSpeechRecognition` because I wouldn't expect them to change after initialization the way a typical state variable would. I certainly wouldn't expect them to change every time `listening` changes, which in my example happens every time I click the button.

I can easily work around this, but I think it would be good to change the behavior of `useSpeechRecognition` instead. It would be more efficient and less surprising if `listen` and `stop` weren't regenerated with every render, or at least if they always referred to the current value of `listening`.

There are a couple of ways to do that:

1. Use `useCallback` and a ref to always get the latest value of `listening` as described in [How to read an often-changing value from useCallback?](https://reactjs.org/docs/hooks-faq.html#how-to-read-an-often-changing-value-from-usecallback)
2. Use `useReducer` and have `listen` and `stop` simply call `dispatch` to update the state. Have a `useEffect` hook that listens for state changes and calls the appropriate methods on the `recognition` object.

I'd be happy to make a pull request if you're interested. Awesome library, by the way!

**Comment by MikeyParton** [comment]
This is all valid, thanks for the feedback. I can take a look at this later when I get some time, but if you'd like to make a pull request, I'd be happy to review it. I'd probably go with the `useCallback` approach

**Comment by maxwellhaydn** [comment]
Ok, here it is: https://github.com/MikeyParton/react-speech-kit/pull/32

**Comment by MikeyParton** [comment]
This is now available in v3.0.1. Thanks again!

**Comment by suhrob008** [comment]
![image](https://github.com/MikeyParton/react-speech-kit/assets/118727649/6add084c-78d8-4bbc-bc69-de58eadb1a0a)

what can you do?

**Comment by suhrob008** [comment]
> This is now available in v3.0.1. Thanks again!

![image](https://github.com/MikeyParton/react-speech-kit/assets/118727649/6add084c-78d8-4bbc-bc69-de58eadb1a0a)

what can you do?

**Comment by suhrob008** [comment]
> > This is now available in v3.0.1. Thanks again!
> 
> ![image](https://private-user-images.githubusercontent.com/118727649/298365604-6add084c-78d8-4bbc-bc69-de58eadb1a0a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU4MjAyNDIsIm5iZiI6MTcwNTgxOTk0MiwicGF0aCI6Ii8xMTg3Mjc2NDkvMjk4MzY1NjA0LTZhZGQwODRjLTc4ZDgtNGJiYy1iYzY5LWRlNThlYWRiMWEwYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEyMVQwNjUyMjJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNmNmOWY5OWQ4OTVmYjA3ZDFmZGQyYjEzNWRhYjgzM2RhOTM0YWZkNDkwZDg2YmM1NDk2MWUxYWFkYzJiMTdjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.PL0gorJ88goNsRZcm1NQ8OMMrI9VgXTc4xmOqZYADN4)
> 
> what can you do?
